File: leetcode/1091_Shortest_Path_in_Binary_Matrix.py
from typing import List
from unittest import TestCase
from collections import deque

class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        n = len(grid)
        dxdy = [(i, j) for i in (-1, 0, 1) for j in (-1, 0, 1) if i != 0 or j != 0]        
        if grid[0][0] == 0:
            queue = deque([(1, 0, 0)])
            while queue:
                cost, x, y = queue.popleft()
                if x == n - 1 and y == n - 1:
                    return cost
                for x1, y1 in ((x + dx, y + dy) for dx, dy in dxdy):
                    if 0 <= x1 < n and 0 <= y1 < n and grid[x1][y1] == 0:
                        queue.append((cost + 1, x1, y1))
                        grid[x1][y1] = 1
        return -1

class ShortestPathBinaryMatrixTest(TestCase):
    
    def test_1(self):
        self.assertEqual(2, Solution().shortestPathBinaryMatrix([[0,1],[1,0]]))
            
    def test_2(self):
        self.assertEqual(4, Solution().shortestPathBinaryMatrix([[0,0,0],[1,1,0],[1,1,0]]))

    def test_3(self):
        self.assertEqual(-1, Solution().shortestPathBinaryMatrix([[1,0,0],[1,1,0],[1,1,0]]))
    

File: leetcode/1091_Shortest_Path_in_Binary_Matrix.rs
use std::collections::VecDeque;
struct Solution {}

impl Solution {
    pub fn shortest_path_binary_matrix(mut grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len() as i32;
        if grid[0][0] == 0 {
            let dxdy: Vec<(i32, i32)> = (-1..=1).into_iter()
                .flat_map(|dx| (-1..=1).into_iter().map(move |dy| (dx, dy)))
                .filter(|(dx, dy)| *dx != 0 || *dy != 0 )
                .collect();
            grid[0][0] = 1;
            let mut queue = VecDeque::from([(1, 0, 0)]);
            while let Some((cost, x, y)) = queue.pop_front() {
                if x == n - 1 && y == n - 1 {
                    return cost
                }
                for (x, y) in  dxdy.iter()
                    .map(|(dx, dy)| (x + dx, y + dy))
                    .filter(|(x, y)|  0 <= *x && *x  < n && 0 <= *y && *y < n) 
                {
                    if grid[x as usize][y as usize] == 0 {
                        queue.push_back((cost + 1, x, y));
                        grid[x as usize][y as usize] = 1;    
                    }
                }
            }
        }
        -1
    }
}

#[test]
fn test_1() {
    assert_eq!(2, Solution::shortest_path_binary_matrix(vec![vec![0,1], vec![1,0]]));
}

#[test]
fn test_2() {
    assert_eq!(4, Solution::shortest_path_binary_matrix(vec![vec![0,0,0], vec![1,1,0], vec![1,1,0]]));
}

#[test]
fn test_3() {
    assert_eq!(-1, Solution::shortest_path_binary_matrix(vec![vec![1,0,0], vec![1,1,0], vec![1,1,0]]));
}
